File: engines/calendar/SolarEngine.js
export class SolarEngine {

    constructor() {
        this.year = 1;
        this.month = 1;
        this.day = 1;
    }


    getDate() {
        return {
            year: this.year,
            month: this.month,
            day: this.day
        };
    }


    nextDay() {

        this.day++;

        let daysInMonth = this.getDaysInMonth(
            this.month,
            this.year
        );


        if (this.day > daysInMonth) {

            this.day = 1;
            this.month++;

        }


        if (this.month > 12) {

            this.month = 1;
            this.year++;

        }

    }


    getDaysInMonth(month, year) {

        const months = [
            31,28,31,30,
            31,30,31,31,
            30,31,30,31
        ];


        let days = months[month - 1];


        if (
            month === 2 &&
            this.isLeapYear(year)
        ) {
            days = 29;
        }


        return days;

    }


    isLeapYear(year) {

        return (
            year % 4 === 0 &&
            (
                year % 100 !== 0 ||
                year % 400 === 0
            )
        );

    }

}

File: GoldenCalendar-Core/src/SolarEngine.cpp
#include "../include/SolarEngine.h"

SolarDate SolarEngine::getDate(long long dayId)
{

    SolarDate date;

    date.year = 1;

    const char* months[] = {
        "يناير",
        "فبراير",
        "مارس",
        "أبريل",
        "مايو",
        "يونيو",
        "يوليو",
        "أغسطس",
        "سبتمبر",
        "أكتوبر",
        "نوفمبر",
        "ديسمبر"
    };

    int daysInMonth[] = {
        31,28,31,30,31,30,
        31,31,30,31,30,31
    };


    long long remaining = dayId;

    int month = 1;


    while(remaining > daysInMonth[month-1])
    {
        remaining -= daysInMonth[month-1];
        month++;
    }


    date.month = month;
    date.day = remaining;
    date.monthName = months[month-1];


    return date;

}
